plugins/synapticsmst/synapticsmst-device.c

```diff
@@ -504,32 +541,502 @@ synapticsmst_device_get_flash_checksum (SynapticsMSTDevice *device,
+	if(rc)								
+		g_debug("Get Bank fail");		
+
+	if((dwData[0] & BIT(7))||(dwData[0] & BIT(30))){
+		BankInUse = BANKTAG_1;
+		BankToUpdate = BANKTAG_0;
+	}
+	#if MSTDBG
+	g_debug("Used %x ToUpd %x",BankInUse, BankToUpdate);
+	#endif
+	/* get firmware size */
+	fw_size =  0x410 + (*(payload_data + 0x400) << 24) + (*(payload_data + 0x401) << 16) + (*(payload_data + 0x402) << 8) + (*(payload_data + 0x403));
+	if( fw_size < payload_len ){
+		fw_size = 104*1024;		// Current firmware size is 104K
+	}
+
+	rc = 0;
+	retryCount = 0;	
+	unitSize = BLOCK_UNIT;//64
+	write_loops = fw_size / unitSize;//(fw_size / unitSize / 6);//Here write 256 byte?
+	if (fw_size % unitSize)
+		write_loops++;
+
+	do{
+		/* erase storage */
+		eraseOffset = BankToUpdate * 2;
+		rc = synapticsmst_device_set_flash_sector_erase(device, FLASH_SECTOR_ERASE_64K, eraseOffset++);
+		if (rc)
+		{
+			g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+						"can't sector erase flash");
+			eraseFail = TRUE;
+			break;
+		}
+		rc = synapticsmst_device_set_flash_sector_erase(device, FLASH_SECTOR_ERASE_64K, eraseOffset);
+		if (rc)
+		{
+			g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+						"can't sector erase flash");
+			eraseFail = TRUE;
+			break;
+		}
+		if(eraseFail)
+			break;
+		/* update */
+		#if MSTDBG
+		g_debug("skip check ESM update");
+		#endif
+		writeIndex = 0;	
+		writeOffset = EEPROM_BANK_OFFSET * BankToUpdate;		
+		
+		for( i = 0; i < write_loops ; i++ )
+		{
+			rc = synapticsmst_common_rc_set_command (connection,
+					 UPDC_WRITE_TO_EEPROM, unitSize, writeOffset, payload_data+ writeIndex);
+			if (rc) {
+				/* repeat once */
+				rc = synapticsmst_common_rc_set_command (connection,
+						 UPDC_WRITE_TO_EEPROM, unitSize, writeOffset, payload_data+ writeIndex);
+			}
+			if (rc)	{
+				g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+							"FW write fail");
+				break;
+			}
+
+			writeOffset += unitSize;
+			writeIndex += unitSize;
+		}	
+
+		/* verify CRC */
+		rc = 0;
+		checksum = synapticsmst_device_get_crc( 0, 16, fw_size, payload_data );
+		for(i=0; i<5; i++)
+		{
+			g_usleep(2000);	
+			if (synapticsmst_common_rc_special_get_command (connection,
+						UPDC_CAL_EEPROM_CHECK_CRC16,
+						fw_size, (EEPROM_BANK_OFFSET * BankToUpdate), NULL, 4, (guint8 *)(&flash_checksum))) {
+				g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+							"Failed to get flash checksum");
+			}
+			if(checksum != flash_checksum){
+				#if MSTDBG
+				g_debug("ck %2x fck %2x", checksum, flash_checksum);
+				#endif
+				rc = 0xFF;
+			}
+			else{
+				rc = 0;
+				break;
+			}
+		}
+
+		if(rc == 0)
+			break;
+		else	
+			g_usleep(3000);	
+	}while(retryCount++ < MAX_RETRY_COUNTS);
+
+	if(retryCount== MAX_RETRY_COUNTS){
+				g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+									"firmware update fail");
+		return FALSE;
+	}
+
+	/* set tag vaild*/	
```

Comment:
Isn't "valid" tag already in the FW?  I would think you shouldn't need to do this part of the write process do you?